sdk-cpp/src/internal/rest/download_impl.cpp

```diff
@@ -24,121 +24,145 @@ namespace deliveryoptimization
 {
 namespace details
 {
-CDownloadImpl::CDownloadImpl(const std::string& uri, const std::string& downloadFilePath)
+int32_t CDownloadImpl::Init(const std::string& uri, const std::string& downloadFilePath) noexcept
 {
-    cpprest_web::uri_builder builder(g_downloadUriPart);
-    builder.append_path("create");
-    builder.append_query("Uri", uri);
-    builder.append_query("DownloadFilePath", downloadFilePath);
-
-    for (int retryAttempts = 0; retryAttempts < g_maxNumRetryAttempts; retryAttempts++)
+    try
     {
-        try
-        {
-            const auto respBody = CHttpClient::GetInstance().SendRequest(HttpRequest::POST, builder.to_string());
-            _id = respBody.get<std::string>("Id");
-            return;
-        }
-        catch (const msdo::exception& e)
+        cpprest_web::uri_builder builder(g_downloadUriPart);
+        builder.append_path("create");
+        builder.append_query("Uri", uri);
+        builder.append_query("DownloadFilePath", downloadFilePath);
+
+        for (int retryAttempts = 0; retryAttempts < g_maxNumRetryAttempts; retryAttempts++)
         {
-            // Handle DOCS in Shutdown state while Create request was issued, client will restart and this loop will break
-            // TODO(jimson): SDK doesn't have the capability to start do-client-lite.service if not running already. Test this
-            // code path when it is implemented.
-            if (e.error_code() != static_cast<int32_t>(msdo::errc::no_service))
+            try
             {
-                throw;
+                const auto respBody = CHttpClient::GetInstance().SendRequest(HttpRequest::POST, builder.to_string());
+                _id = respBody.get<std::string>("Id");
+                return S_OK;
             }
-            if (retryAttempts < g_maxNumRetryAttempts - 1)
+            catch (const msdo::exception& e)
             {
-                std::this_thread::sleep_for(1s);
+                // Handle DOCS in Shutdown state while Create request was issued, client will restart and this loop will break
+                // TODO(jimson): SDK doesn't have the capability to start do-client-lite.service if not running already. Test this
+                // code path when it is implemented.
+                if (e.error_code() != static_cast<int32_t>(msdo::errc::no_service))
+                {
+                    return e.error_code();
+                }
+                if (retryAttempts < g_maxNumRetryAttempts - 1)
+                {
+                    std::this_thread::sleep_for(1s);
+                }
             }
         }
+        return static_cast<int32_t>(msdo::errc::no_service);
+    }
+    catch (msdo::exception& e)
```

Comment:
>msdo::exception [](http://example.com/codeflow?start=11&length=15)

All invoked methods here throw only this exception type?

Reply:
Actually the uri builder invoked methods don't throw exceptions as far as I can tell.